File: ios/GradeThread/Inventory/ItemCanvas/ItemDraft.swift
import Foundation

/// Editable snapshot of an `inventory_items` row used by ItemCanvasView.
/// Value type so the canvas can compare a draft against the original
/// snapshot to compute `isDirty` without holding two SwiftData instances.
public struct ItemDraft: Equatable {
    public var title: String
    public var brand: String
    public var sku: String
    public var size: String
    public var color: String
    public var material: String
    public var conditionNotes: String
    public var status: String
    public var category: FlipdeskCategory?
    /// Clothing classification raw values ("" = unset). Required to grade a
    /// clothing item; only edited when `category == .clothing`.
    public var garmentType: String
    public var garmentCategory: String
    /// Buyer-facing listing copy, style/variant note, and who sourced the item
    /// (web parity). Free-form; empty saves as NULL.
    public var itemDescription: String
    public var style: String
    public var sourcedBy: String
    /// Acquisition date (nil = unset), storage/acquisition container label, and
    /// the hand-curated comparable sales (web parity).
    public var acquiredDate: Date?
    public var container: String
    public var compSet: [ItemComp]
    public var targetPriceText: String
    public var acquiredPriceText: String
    /// US-676: storage location / bin label and consignment link.
    public var locationBin: String
    public var consignorId: String?
    /// Per-item consignor split override as a 0–100 string ("" = use the
    /// consignor's default).
    public var consignmentSplitText: String
    /// Editable flat garment measurements keyed by canonical key (see
    /// ``MeasurementCatalog``). LENGTH values are inches. Edited in the canvas
    /// Measurements section; encoded to the `measurements` jsonb column on save.
    public var measurements: [String: Double]

    public init(
        title: String = "",
        brand: String = "",
        sku: String = "",
        size: String = "",
        color: String = "",
        material: String = "",
        conditionNotes: String = "",
        status: String = "cataloged",
        category: FlipdeskCategory? = nil,
        garmentType: String = "",
        garmentCategory: String = "",
        itemDescription: String = "",
        style: String = "",
        sourcedBy: String = "",
        acquiredDate: Date? = nil,
        container: String = "",
        compSet: [ItemComp] = [],
        targetPriceText: String = "",
        acquiredPriceText: String = "",
        locationBin: String = "",
        consignorId: String? = nil,
        consignmentSplitText: String = "",
        measurements: [String: Double] = [:]
    ) {
        self.title = title
        self.brand = brand
        self.sku = sku
        self.size = size
        self.color = color
        self.material = material
        self.conditionNotes = conditionNotes
        self.status = status
        self.category = category
        self.garmentType = garmentType
        self.garmentCategory = garmentCategory
        self.itemDescription = itemDescription
        self.style = style
        self.sourcedBy = sourcedBy
        self.acquiredDate = acquiredDate
        self.container = container
        self.compSet = compSet
        self.targetPriceText = targetPriceText
        self.acquiredPriceText = acquiredPriceText
        self.locationBin = locationBin
        self.consignorId = consignorId
        self.consignmentSplitText = consignmentSplitText
        self.measurements = measurements
    }

    /// Populate from a LocalInventoryItem. The numeric fields go through
    /// a string box so the user can type freely; CurrencyFormatter
    /// parses back to Double on save.
    init(from item: LocalInventoryItem, currencyFormatter: CurrencyFormatter = CurrencyFormatter()) {
        self.title = item.title
        self.brand = item.brand ?? ""
        self.sku = item.sku ?? ""
        self.size = item.size ?? ""
        self.color = item.color ?? ""
        self.material = item.material ?? ""
        self.conditionNotes = item.conditionNotes ?? ""
        self.status = item.status
        // Seed the category from the synced `item_category` so the picker shows
        // the real value (and a save round-trips it instead of nulling it out),
        // and so the clothing-only garment pickers gate correctly.
        self.category = item.itemCategory.flatMap(FlipdeskCategory.init(rawValue:))
        self.garmentType = item.garmentType ?? ""
        self.garmentCategory = item.garmentCategory ?? ""
        self.itemDescription = item.itemDescription ?? ""
        self.style = item.style ?? ""
        self.sourcedBy = item.sourcedBy ?? ""
        self.acquiredDate = item.acquiredDate
        self.container = item.container ?? ""
        self.compSet = ItemComp.decodeList(item.compSetJSON)
        self.targetPriceText = item.targetPrice.map { currencyFormatter.formatRaw($0) } ?? ""
        self.acquiredPriceText = item.acquiredPrice.map { currencyFormatter.formatRaw($0) } ?? ""
        self.locationBin = item.locationBin ?? ""
        self.consignorId = item.consignorId
        // US-1491: seed locale-formatted so the field's display matches the
        // locale-aware parser (parseSplit). A "."-formatted seed re-parses wrong
        // in comma-decimal locales.
        self.consignmentSplitText = item.consignmentSplitPct
            .map { CurrencyFormatter().formatRaw($0) } ?? ""
        self.measurements = ItemCanvasView.decodeMeasurements(item.measurementsJSON) ?? [:]
    }
}

/// Status transition guard. Mirrors the web's `resolveStatus` policy: a
/// row that's already reached a "terminal" sale state must not regress
/// to a pre-sale status by accident.
public enum StatusGuard {
    /// Terminal states an item shouldn't be silently demoted from.
    public static let terminalStates: Set<String> = [
        "sold", "shipped", "completed", "returned", "archived",
    ]

    /// Pre-sale states. Defined by exclusion: anything not terminal
    /// counts as "pre-sale".
    public static func isPreSale(_ status: String) -> Bool {
        !terminalStates.contains(status)
    }

    /// Returns true iff a transition from `current` to `next` is allowed.
    /// Once a row is in a terminal state, only other terminal states are
    /// reachable — you can flip sold → shipped → completed, but you
    /// can't reset back to cataloged without an explicit override
    /// (which we leave to a future admin tool).
    public static func allows(from current: String, to next: String) -> Bool {
        if current == next { return true }
        if terminalStates.contains(current), !terminalStates.contains(next) {
            return false
        }
        return true
    }
}

File: android/app/src/main/java/com/gradethread/app/inventory/ItemDraft.kt
package com.gradethread.app.inventory

import com.gradethread.app.capture.CurrencyAmount
import com.gradethread.app.capture.FlipdeskCategory
import com.gradethread.app.sync.db.InventoryItemEntity
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * US-1343: the editable snapshot of an `inventory_items` row.
 *
 * A value type, so the canvas can diff a draft against the snapshot it was
 * seeded from without holding two live Room entities. Everything numeric is
 * boxed as text while editing — a seller typing "12." must not be reinterpreted
 * mid-keystroke.
 */
data class ItemDraft(
    val title: String = "",
    val brand: String = "",
    val sku: String = "",
    val size: String = "",
    val color: String = "",
    val material: String = "",
    val style: String = "",
    val status: String = "cataloged",
    /**
     * NULLABLE on purpose. `FlipdeskCategory.from()` falls back to CLOTHING for
     * an unknown value, which is right for a picker default and WRONG here: a
     * photo-first item is created with no category at all, and seeding it to
     * clothing would have the first save silently assert a category the seller
     * never chose — on a watch, say.
     */
    val category: FlipdeskCategory? = null,
    val garmentType: String = "",
    val garmentCategory: String = "",
    val description: String = "",
    val conditionNotes: String = "",
    val sourcedBy: String = "",
    val container: String = "",
    val locationBin: String = "",
    /** Epoch millis; a Postgres DATE, so only the calendar day is written. */
    val acquiredDate: Long? = null,
    val acquiredPriceText: String = "",
    val targetPriceText: String = "",
    val consignorId: String? = null,
    /** 0–100 as typed; "" means "use the consignor's default". */
    val consignmentSplitText: String = "",
    /** US-1345: canonical key → value, decoded from the `measurements` jsonb. */
    val measurements: Map<String, Double> = emptyMap(),
    /** US-1346: the seller's own saved comparable sales (`comp_set` jsonb). */
    val comps: List<ItemComp> = emptyList(),
    /** US-1347: the resolved eBay leaf category for this item's specifics. */
    val ebayCategoryId: String? = null,
    /** US-1347: `ebay_aspects` — aspect name → values. */
    val aspects: Map<String, List<String>> = emptyMap(),
    /** US-1347: the parallel provenance map (00184). */
    val aspectSources: Map<String, AspectSync.Provenance> = emptyMap(),
) {

    /** Garment classification only applies to clothing. */
    val showsGarmentFields: Boolean get() = category == FlipdeskCategory.CLOTHING

    companion object {

        fun from(item: InventoryItemEntity): ItemDraft = ItemDraft(
            title = item.title,
            brand = item.brand.orEmpty(),
            sku = item.sku.orEmpty(),
            size = item.size.orEmpty(),
            color = item.color.orEmpty(),
            material = item.material.orEmpty(),
            style = item.style.orEmpty(),
            status = item.status,
            // Nullable seed — see the property doc.
            category = categoryOrNull(item.itemCategory),
            garmentType = item.garmentType.orEmpty(),
            garmentCategory = item.garmentCategory.orEmpty(),
            description = item.itemDescription.orEmpty(),
            conditionNotes = item.conditionNotes.orEmpty(),
            sourcedBy = item.sourcedBy.orEmpty(),
            container = item.container.orEmpty(),
            locationBin = item.locationBin.orEmpty(),
            acquiredDate = item.acquiredDate,
            acquiredPriceText = CurrencyAmount.formatRaw(item.acquiredPrice?.let { toCents(it) }),
            targetPriceText = CurrencyAmount.formatRaw(item.targetPrice?.let { toCents(it) }),
            consignorId = item.consignorId,
            consignmentSplitText = item.consignmentSplitPct
                ?.let { CurrencyAmount.formatRaw(toCents(it)) }
                .orEmpty(),
            measurements = MeasurementCatalog.decode(item.measurementsJson),
            comps = CompSet.decode(item.compSetJson),
            ebayCategoryId = item.ebayCategoryId,
            aspects = AspectSync.decodeAspects(item.ebayAspectsJson),
            aspectSources = AspectSync.decodeSources(item.ebayAspectSourcesJson),
        )

        /** Strict lookup: an absent or unrecognized value stays null. */
        fun categoryOrNull(wire: String?): FlipdeskCategory? =
            wire?.takeIf { it.isNotBlank() }
                ?.let { value -> FlipdeskCategory.entries.firstOrNull { it.wire == value } }

        private fun toCents(value: Double): Long = Math.round(value * 100)

        /** `YYYY-MM-DD` — a Postgres DATE, never a timestamp. */
        fun toDateWire(epochMillis: Long?): String? = epochMillis?.let {
            java.time.Instant.ofEpochMilli(it)
                .atZone(java.time.ZoneId.systemDefault())
                .toLocalDate()
                .toString()
        }
    }
}

/**
 * Turning an edited draft into a write.
 *
 * Kept separate and pure because two rules here are easy to get wrong and
 * expensive when they are.
 */
object ItemPatch {

    /**
     * Terminal states an item must not silently regress from. A row that has
     * sold cannot go back to `cataloged` because a stale canvas was open.
     */
    val TERMINAL_STATES = setOf("sold", "shipped", "completed", "returned", "archived")

    fun allowsStatus(current: String, next: String): Boolean =
        current == next || !(current in TERMINAL_STATES && next !in TERMINAL_STATES)

    /**
     * The columns that actually CHANGED between [original] and [edited].
     *
     * Only changed keys are emitted, which matters more than it looks: the
     * canvas holds a snapshot taken when it opened, and a full-row write would
     * push every stale field back over anything that changed in between — a
     * grade that landed, a price a sync pulled, a field another device edited.
     * This is the same class of bug as the publish-time listing snapshots.
     *
     * A field cleared to blank becomes JSON null, never "": an empty string is
     * a real value that reads as set-but-empty downstream, and it would occupy
     * the partial unique index on `sku`.
     */
    fun diff(original: ItemDraft, edited: ItemDraft): JsonObject {
        val patch = mutableMapOf<String, JsonElement>()

        fun text(column: String, before: String, after: String) {
            if (before.trim() == after.trim()) return
            patch[column] = after.trim().takeIf { it.isNotEmpty() }
                ?.let { JsonPrimitive(it) } ?: JsonNull
        }

        text("title", original.title, edited.title)
        text("brand", original.brand, edited.brand)
        text("sku", original.sku, edited.sku)
        text("size", original.size, edited.size)
        text("color", original.color, edited.color)
        text("material", original.material, edited.material)
        text("style", original.style, edited.style)
        text("garment_type", original.garmentType, edited.garmentType)
        text("garment_category", original.garmentCategory, edited.garmentCategory)
        text("description", original.description, edited.description)
        text("condition_notes", original.conditionNotes, edited.conditionNotes)
        text("sourced_by", original.sourcedBy, edited.sourcedBy)
        text("container", original.container, edited.container)
        text("location_bin", original.locationBin, edited.locationBin)

        // Title is NOT NULL server-side, so a blank title is a no-op rather
        // than a write that fails the whole patch.
        if (patch["title"] is JsonNull) patch.remove("title")

        if (original.category != edited.category) {
            // Only written when it actually changed, so an item that never had
            // a category keeps not having one (AC3).
            patch["item_category"] = edited.category?.let { JsonPrimitive(it.wire) } ?: JsonNull
        }

        if (original.status != edited.status && allowsStatus(original.status, edited.status)) {
            patch["status"] = JsonPrimitive(edited.status)
        }

        if (original.acquiredDate != edited.acquiredDate) {
            patch["acquired_date"] = ItemDraft.toDateWire(edited.acquiredDate)
                ?.let { JsonPrimitive(it) } ?: JsonNull
        }

        money("acquired_price", original.acquiredPriceText, edited.acquiredPriceText, patch)
        money("target_price", original.targetPriceText, edited.targetPriceText, patch)
        money(
            "consignment_split_pct",
            original.consignmentSplitText,
            edited.consignmentSplitText,
            patch,
        )

        if (original.measurements != edited.measurements) {
            // The whole document is replaced, because a jsonb UPDATE replaces
            // it anyway — sending only the changed keys would silently delete
            // every measurement the seller did not touch this session.
            patch["measurements"] = MeasurementCatalog.encode(edited.measurements)
                ?.let { kotlinx.serialization.json.Json.parseToJsonElement(it) }
                ?: JsonNull
        }

        if (original.comps != edited.comps) {
            // Whole-document replace, same rule as measurements: a jsonb UPDATE
            // replaces it anyway, so a partial write would drop the comps the
            // seller didn't touch.
            patch["comp_set"] = CompSet.encode(edited.comps)
                ?.let { kotlinx.serialization.json.Json.parseToJsonElement(it) }
                ?: JsonNull
        }

        // US-1347: the columns own their aspects, so the projection runs on
        // EVERY save — otherwise a brand edit here and the Brand specific that
        // publishes to eBay drift apart silently.
        // Compared against the ORIGINAL'S PROJECTION, not against its stored
        // aspects. An item can carry a brand with no Brand specific yet, and
        // comparing to the stored map would make that difference show up on
        // every open — leaving the canvas permanently "dirty" and writing
        // aspects nobody edited. Comparing projection to projection means the
        // backfill rides along with a real edit instead.
        val (originalProjected, originalSources) = AspectSync.projectColumnAspects(
            original,
            original.aspects,
            original.aspectSources,
        )
        val (projected, projectedSources) = AspectSync.projectColumnAspects(
            edited,
            edited.aspects,
            edited.aspectSources,
        )
        val prunedSources = AspectSync.pruneSources(projectedSources, projected)
        val originalPruned = AspectSync.pruneSources(originalSources, originalProjected)
        if (projected != originalProjected) {
            patch["ebay_aspects"] = AspectSync.encodeAspects(projected)
                ?.let { kotlinx.serialization.json.Json.parseToJsonElement(it) }
                ?: JsonNull
        }
        if (prunedSources != originalPruned) {
            patch["ebay_aspect_sources"] = AspectSync.encodeSources(prunedSources)
                ?.let { kotlinx.serialization.json.Json.parseToJsonElement(it) }
                ?: JsonNull
        }
        if (original.ebayCategoryId != edited.ebayCategoryId) {
            patch["ebay_category_id"] = edited.ebayCategoryId?.takeIf { it.isNotBlank() }
                ?.let { JsonPrimitive(it) } ?: JsonNull
        }

        if (original.consignorId != edited.consignorId) {
            patch["consignor_id"] = edited.consignorId?.takeIf { it.isNotBlank() }
                ?.let { JsonPrimitive(it) } ?: JsonNull
        }

        return JsonObject(patch)
    }

    /**
     * Money compares as CENTS, so "12" and "12.00" are not a change.
     *
     * Without this every re-open of the canvas would dirty the price fields
     * purely by reformatting them, and every save would write values nobody
     * touched.
     */
    private fun money(
        column: String,
        before: String,
        after: String,
        patch: MutableMap<String, JsonElement>,
    ) {
        val beforeCents = CurrencyAmount.parseCents(before)
        val afterCents = CurrencyAmount.parseCents(after)
        if (beforeCents == afterCents) return
        patch[column] = afterCents
            ?.let { JsonPrimitive(CurrencyAmount.formatRaw(it).toDouble()) }
            ?: JsonNull
    }

    /** Whether anything at all would be written. */
    fun isDirty(original: ItemDraft, edited: ItemDraft): Boolean =
        diff(original, edited).isNotEmpty()
}
